backend/src/main/scala/cromwell/backend/validation/ValidatedRuntimeAttributesBuilder.scala

```diff
@@ -0,0 +1,90 @@
+
+trait ValidatedRuntimeAttributesBuilder {
+
+  /**
+    * Returns the validations that should be used and returned during evaluation.
+    *
+    * These validations will be checked on backend initialization against all the calls. If a validation value will not
+    * be supported, but should be checked anyway, supply it in `valuesOnlyValidations`, not here..
+    *
+    * @return the validations that should be used and returned during evaluation.
+    */
+  def validations: Seq[RuntimeAttributesValidation[_]]
+
+  /**
+    * Returns the additional validations that should be used during value parsing.
+    *
+    * For example, sometimes docker might not be supported, BUT we want to still validate the value if specified.
+    *
+    * In that case, return the validation here.
+    *
+    * @return the additional validations that should be used during value parsing.
+    */
+  protected def validationsInternal: Seq[RuntimeAttributesValidation[_]] = Seq.empty
+
+  def unsupportedKeys(keys: Seq[String]): Seq[String] = keys.diff(validationKeys)
+
+  private lazy val coercionMap: Map[String, Traversable[WdlType]] =
+    validations.map(attr => attr.key -> attr.coercion).toMap
+
+  private lazy val validationKeys = validations.map(_.key)
+
+  private lazy val staticDefaults: Map[String, WdlValue] = (validations collect {
+    case attr if attr.staticDefaultOption.isDefined => attr.key -> attr.staticDefaultOption.get
+  }).toMap
+
+  def build(attrs: Map[String, WdlValue], options: WorkflowOptions, logger: Logger): ValidatedRuntimeAttributes = {
+    import RuntimeAttributesDefault._
+
+    // Fail now if some workflow options are specified but can't be parsed correctly
+    val defaultFromOptions = workflowOptionsDefault(options, coercionMap).get
+    val withDefaultValues: Map[String, WdlValue] = withDefaults(attrs, List(defaultFromOptions, staticDefaults))
+
+    RuntimeAttributesValidation.warnUnrecognized(withDefaultValues.keySet, validationKeys.toSet, logger)
+
+    val runtimeAttributesErrorOr: ErrorOr[ValidatedRuntimeAttributes] = validate(withDefaultValues)
+    runtimeAttributesErrorOr match {
+      case Success(runtimeAttributes) => runtimeAttributes
+      case Failure(nel) => throw new RuntimeException with MessageAggregation {
+        override def exceptionContext: String = "Runtime attribute validation failed"
+
+        override def errorMessages: Traversable[String] = nel.list
+      }
+    }
+  }
+
+  private def validate(values: Map[String, WdlValue]): ErrorOr[ValidatedRuntimeAttributes] = {
+    val validationsForValues = validations ++ validationsInternal
+    val errorsOrValuesMap = validationsForValues.map(validation => validation.key -> validation.validate(values))
+
+    import scalaz.Scalaz._
+    import scalaz._
+
+    implicit val semigroupForMap = new Semigroup[Map[String, Any]] {
+      override def append(f1: Map[String, Any], f2: => Map[String, Any]) = f1 ++ f2
+    }
```

Comment:
This is actually not a `Semigroup`.  No `Semigroup` is asserted for the value type so it's not clear how merges should be handled.  More info at

http://stackoverflow.com/questions/7076128/best-way-to-merge-two-maps-and-sum-the-values-of-same-key